---
title: "EnrichDO: a Global Weighted Model for Disease Ontology Enrichment Analysis"
author: 
- name: Liang Cheng
  affiliation: College of Bioinformatics Science and Technology, Harbin Medical University
- name: Haixiu Yang
  email: yanghaixiu@ems.hrbmu.edu.cn
  affiliation: College of Bioinformatics Science and Technology, Harbin Medical University
- name: Hongyu Fu
  affiliation: College of Bioinformatics Science and Technology, Harbin Medical University
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
package: EnrichDO
vignette: >
  %\VignetteIndexEntry{EnrichDO: Disease Ontology Enrichment Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Disease Ontology (DO) enrichment analysis is an effective means to discover the associations between genes and diseases. However, most current DO-based enrichment methods were unable to solve the over enriched problem caused by the “true-path” rule. To address this problem, we presented EnrichDO, a double weighted iterative model by integrating the DO graph topology on a global scale. EnrichDO was based on the latest annotations of the human genome with DO terms, and double weighted the annotated genes. On one hand, to reinforce the saliency of direct gene-DO annotations, different initial weights were assigned to directly annotated genes and indirectly annotated genes, respectively. On the other hand, to detect locally most significant node between the parent and its children, less significant nodes were dynamically down-weighted. EnrichDO exhibits higher accuracy that often yield more specific significant DO terms, which alleviate the over enriched problem.

EnrichDO encompasses a variety of statistical models and visualization schemes for discovering the disease-gene relationships under biological big data. Currently uploaded to Bioconductor, we anticipate that our R package will provide a more convenient and effective DO enrichment tool.

```{r setup,results='hide'}
library(EnrichDO)
```

# Weighted DO Enrichment Analysis

EnrichDO is a double weighted iterative model for DO enrichment analysis. Based on the latest annotations of the human genome with DO terms, EnrichDO can identify locally significant enriched nodes by applying different initial weights and dynamic weights for annotated genes and integrating the DO graph topology on a global scale. EnrichDO is an effective and flexible model that supplies various statistical testing models and multiple testing correction methods.

## doEnrich function

In EnrichDO, we implemented ***doEnrich*** to realize the enrichment analysis of ontology by combining topological properties of ontology graph structure.

### Parameter introduction

***doEnrich*** has ten parameters, details:

-   *interestGenes* is the interesting protein coding gene set in the form of Entrez ID.

-   *test* sets the statistical model , which can be "fisherTest", "hypergeomTest", "binomTest", "chisqTest" and "logoddTest" (default is "hypergeomTest").

-   *method* sets the multiple testing correction method, which can be "holm","hochberg", "hommel", "bonferroni", "BH", "BY","fdr" and "

-   *m* sets the maximum number of ancestor layers for ontology enrichment (default is layer 1).

-   *maxGsize* and *minGsize* set the maximum and minimum number of annotated genes of DO terms. DO terms with annotated genes more than *maxGsize* (and less than *minGsize*) are ignored, and the p-value of these DO terms are set to 1 (default *maxGsize* is 5000, *minGsize* is 5).

-   *traditional* is a logical parameter, TRUE for classic ORA enrichment analysis, FALSE for weighted enrichment analysis (Default is FALSE).

-   *delta* sets the threshold of nodes. If the p-value of the DO term is greater than delta, the node is not significant and is not dynamically weighted (Default is 0.01).

-   *result_do* writes the results out to a file by the writeResult function, which is used to visualize the enrichment results (without running the enrichment operation again). Default is NULL.

-   *penalize* is a logical value, and if TRUE, the algorithm reduces the weight again for nodes that are less significant. Default is TRUE.

### Result description

In the following example, several genes (*demo.data*) are randomly selected from the protein-coding genes for analysis. The parameters of ***doEnrich*** is default.

```{r label=init,eval=TRUE,echo=TRUE,collapse=FALSE,results='hide',cache=TRUE}
demo.data=c(1636,351,102,2932,3077,348,4137,54209,5663,5328,23621,3416,3553)
demo_result<-doEnrich(interestGenes = demo.data, 
                      test         = "hypergeomTest", 
                      method       = "BH", 
                      m            = 1, 
                      maxGsize     = 5000,
                      minGsize     = 5,
                      traditional  = FALSE,
                      delta        = 0.01, 
                      result_do    = NULL,
                      penalize     = TRUE)

# [1] "Descending rights test"
# LEVEL: 13	1 nodes	72 genes to be scored
# LEVEL: 12	2 nodes	457 genes to be scored
# LEVEL: 11	3 nodes	907 genes to be scored
# LEVEL: 10	13 nodes	2279 genes to be scored
# LEVEL: 9	54 nodes	6504 genes to be scored
# LEVEL: 8	130 nodes	9483 genes to be scored
# LEVEL: 7	198 nodes	11209 genes to be scored
# LEVEL: 6	220 nodes	12574 genes to be scored
# LEVEL: 5	198 nodes	12936 genes to be scored
# LEVEL: 4	103 nodes	12824 genes to be scored
# LEVEL: 3	30 nodes	11683 genes to be scored
# LEVEL: 2	5 nodes	8032 genes to be scored
# LEVEL: 1	0 nodes	0 genes to be scored
# [1] "BH"
# [1] "hypergeomTest"
```

From the above output results, we can observe the nodes and total genes involved in each layer of DAG structure, as well as the enrichment analysis method and statistical test model used.

The results of the weighted enrichment analysis algorithm are as follows:

```{r eval=TRUE}
head(demo_result)
```

The result of ***doEnrich*** consists of data frame *enrich* and *doterms,* which have been written into environment variables. There are 16 columns of *enrich*, including:

-   the DOterm ID on enrichment (*DOID*),

-   the hierarchy of the DOterm in the DAG graph (*level*),

-   all genes related to the DOterm (*gene.arr*),

-   gene weights in each node (*weight.arr*),

-   the parent node of the DOterm (*parent.arr*) and its number (*parent.len*).

-   child nodes of the DOterm (*child.arr*) and its number (*child.len*),

-   the number of all genes related to the DOterm (*gene.len*),

-   the standard name of the DOterm (*DOTerm*),

-   the weight of annotated genes (*gene.w*),

-   the P-value of the DOterm (*p*), which arrange the order of enrich, and the value of P-value correction (*p.adjust*),

-   the genes of interest annotated to this DOterm (*cg.arr*) and its number (*cg.len*),

-   the number of genes in the interest gene set (*ig.len*).

The data frame *doterms* contains the information of the disease ontology for DAG construction. *doterms* has ten columns including *DOID, level, gene.arr, weight.arr, parent.arr, parent.len, child.arr, child.len, gene.len,* and *DOTerm*.

```{r eval=TRUE}
head(doterms)
```

The default enrichment result of the ***doEnrich*** function is stored in "*enrich*". However, in order to avoid confusion, we do not recommend using *enrich* directly when multiple enrichment operations are required.

```{r eval=FALSE}
doEnrich(interestGenes = demo.data)
head(enrich)
```

### Application cases of doEnrich function

1.Weighted enrichment analysis with multiple parameters. Each parameter in the following example is suitable for enrichment analysis with weights.

```{r eval=FALSE}
weighted_demo<-doEnrich(interestGenes= demo.data,
                           test         = "hypergeomTest",
                           method       = "holm",
                           m            = 1,
                           minGsize     = 5,
                           maxGsize     = 500,
                           delta        = 0.01,
                           penalize     = TRUE)

```

2.The parameter penalize was used to alleviate the impact of different magnitudes of p-values, default value is TRUE. When set to false, the degree of reduction in weight for non-significant nodes is decreased, resulting in a slight increase in significance for these nodes, i.e., their p-value will be reduced.

```{r eval=FALSE}
penalF_demo<-doEnrich(interestGenes = demo.data, penalize = FALSE)

```

3.Using the traditional enrichment analysis method, it doesn't reduce weights according to the DAG structure. Parameters *test, method, m, maxGsize* and *minGsize* can be used flexibly.

```{r eval=FALSE}
Tradition_demo<-doEnrich(demo.data , traditional = TRUE)

# [1] "Traditional test"
# [1] "BH"
# [1] "hypergeomTest"
```

## writeDoTerms function

***writeDoTerms*** can output *DOID, DOTerm, level, genes, parents, children, gene.len, parent.len* and *child.len* in the data frame *doterms* as text. The default file name is "doterms.txt".

```{r eval=TRUE}
writeDoTerms(doterms,file = file.path(tempdir(),"doterms.txt"))
```

## writeResult function

The ***writeResult*** function can output *DOID, DOTerm, p, p.adjust, geneRatio, bgRatio* and *cg* in the data frame *enrich* as text. The default file name is "result.txt".

*geneRatio* represents the intersection of the doterm with the interest set divided by the interest gene set, and *bgRatio* represents all genes of the doterm divided by the background gene set.

***writeResult*** has four parameters. *enrich* indicates the enrichment result of ***doEnrich***, *file* indicates the write address of a file. The parameter *Q* (and *P*) indicates that doterm is output only when *p.adjust* (and *p* value) is less than or equal to *Q* (and *P*). The default values for *P* and *Q* are 1.

```{r eval=TRUE}
writeResult(demo_result,file = file.path(tempdir(),"result.txt"),Q=1,P=1)
```

# Visualization of enrichment results

EnrichDO provides four methods to visualize enrichment results, including bar plot (***drawBarGraph***), bubble plot (***drawPointGraph***), tree plot (***drawGraphviz***) and heatmap (***drawHeatmap***), which can show the research results more concisely and intuitively. Pay attention to the threshold setting for each visual method, if the threshold is too low, the display is insufficient.

## drawBarGraph function

***drawBarGraph*** can draw the top *n* nodes with the most significant p-value as bar chart, and the node's p-value is less than *delta* (By default, *n* is 10 and *delta* is 1e-15).

```{r fig.cap="bar plot",fig.align='center',fig.width=7,fig.height=5}
drawBarGraph(demo_result,n=10,delta = 0.05)
```

## drawPointGraph function

***drawPointGraph*** can draw the top *n* nodes with the most significant p-value as bubble plot, and the node's p-value is less than *delta* (By default, *n* is 10 and *delta* is 1e-15).

```{r fig.cap="point plot",fig.align='center',fig.width=7,fig.height=5}
drawPointGraph(demo_result,n=10,delta = 0.05)
```

## drawGraphViz function

***drawGraphViz*** draws the DAG structure of the most significant *n* nodes, and *labelfontsize* can set the font size of labels in nodes (By default, *n* is 10 and *labelfontsize* is 14). The characters in the figure are the doterm's name corresponding to each node .

In addition, the ***drawGraphViz*** function can also display the P-value of each node in the enrichment analysis (*pview*=TRUE), and the number of overlapping genes of each doterm and interest set (*numview*=TRUE).

```{r fig.cap="tree plot",fig.align='center',fig.width=7,fig.height=5}

drawGraphViz(demo_result, n=10, numview=FALSE, pview=FALSE,labelfontsize = 17)

```

## drawHeatmap function

***drawHeatmap*** function visualizes the strength of the relationship between the top *DOID_n* nodes from enrichment results and the genes whose weight sum ranks the top *gene_n* in these nodes. And the gene displayed must be included in the gene of interest. *readable* indicates whether the gene is displayed as its symbol.

***drawHeatmap*** also provides additional parameters from the pheatmap function, which you can set according to your needs. Default *DOID_n* is10, *gene_n* is 50, *fontsize_row* is 10, *readable* is TRUE.

```{r fig.cap="heatmap",fig.align='center',fig.width=7,fig.height=5}
drawHeatmap(interestGenes=demo.data,
            enrich = demo_result,
            gene_n = 10,
            fontsize_row = 8,
            readable=TRUE)
```

## convenient drawing

Draw(***drawBarGraph ,drawPointGraph ,drawGraphViz***) from ***writeResult*** output files, so you don't have to wait for the algorithm to run.

```{r eval=FALSE}
#Firstly, read the wrireResult output file,using the following two lines
data<-read.delim(file.path(system.file("examples", package = "EnrichDO"),"result.txt"))
doEnrich(result_do = data)

#then, Use the drawing function you need
drawGraphViz(enrich)    #Tree diagram
drawPointGraph(enrich)  #Bubble diagram
drawBarGraph(enrich)    #Bar plot
```
